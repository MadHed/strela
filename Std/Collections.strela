module Std.Collections {
    /**
     * Contiguous storage resizable list
     */
    export class List<T> {
        var array: T[];
        var length: int;

        function init() {
            this.array = new T[](1);
            this.length = 0;
        }

        function append(value: T) {
            var newArray = new T[](this.array.length + 1);
            var i = 0;
            while(i < this.array.length) {
                newArray[i] = this.array[i];
                i = i + 1;
            }
            this.array = newArray;
            this.array[this.length] = value;
            this.length = this.length + 1;
        }

        function set(index: int, value: T) {
            this.array[index] = value;
        }

        function [](i: int): T {
            return this.array[i];
        }
    }

    /**
     * Key-Value storage
     */
    export class Map<K, V> {
        var keys: List<K>;
        var values: List<V>;

        function init() {
            this.keys = new List<K>;
            this.values = new List<V>;
        }

        function has(key: K): bool {
            return this.keyIndex(key) >= 0;
        }

        function keyIndex(key: K): i64 {
            var i: u64 = 0;
            while (i < this.keys.length) {
                if (this.keys[i] == key) return i;
                i += 1;
            }
            return -1;
        }

        function set(key: K, val: V) {
            var idx = this.keyIndex(k);
            if (idx >= 0) {
                this.values[idx] = v;
            }
            else {
                this.keys.append(k);
                this.values.append(v);
            }
        }

        function [](key: K): V {
            return this.values(this.keyIndex(kex));
        }
    }
}