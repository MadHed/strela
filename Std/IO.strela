// Copyright (c) 2018 Stephan Unverwerth
// This code is licensed under MIT license (See LICENSE for details)

module Std.IO {
    export function print(v: int): void {}
    export function print(v: float): void {}
    export function print(v: String): void {}
    export function print(v: bool): void {}

    export function println(v: int): void { print(v); print("\n"); }
    export function println(v: float): void { print(v); print("\n"); }
    export function println(v: String): void { print(v); print("\n"); }
    export function println(v: bool): void { print(v); print("\n"); }
}