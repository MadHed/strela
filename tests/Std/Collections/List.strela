// Copyright (c) 2018 Stephan Unverwerth
// This code is licensed under MIT license (See LICENSE for details)

module Tests.Collections.List {
    import Std.IO.println;
    import Std.Collections.List;

    function main(args: String[]): int {
        var list = new List<int>;
        list.append(1);
        list.append(2);
        println(list.length);
        println(list[0]);
        println(list[1]);

        var list2: List<String> = ["A", "B"];
        println(list2[0]);
        println(list2[1]);

        list2.remove(0);
        list2.remove(1);

        println(list2.length);
        return 0;
    }
}