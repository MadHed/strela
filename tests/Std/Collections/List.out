2
1
2
A
B
0
